### DISCUSSION TITLE
Is there a way to get `int`-index that corresponds to a given `Index`?

### DISCUSSION THREAD

**Discussion by ArtemyB** [fsprojects/FSharp.Data.Adaptive#107] [Q&A] [answered]
Hi, in the task I'm solving currently I need to get an int-index for `alist<'T>`'s element (or for the `Index` of the element). I see there is `tryAt: int -> Index option` function in `AList` module, however couldn't find anything like a reverse one (i.e. `Index -> int option`). Is there an "easy" way to do that at all, or the only option is to create some kind of separate `Index`-`int` cache (maybe by folding the list to `amap<Index, int>` -- not sure whether it's a good solution)?

**Comment by krauthaufen** [comment]
Hey, the main problem here is, that the `int` may actually change when something is inserted into the list.
That is why we can't really give you an `int` for an element.

It would however be possible to implement something like `AList.mapIntIndex : (aval<int> -> 'a -> 'b) -> alist<'a> -> alist<'b>` where the `aval<int>` will change appropriately. Of course for small lists you can always do something like `list |> AList.toAVal |> AVal.map (Seq.mapi mapping) |> AList.ofAVal` but that looses adaptivity on its elements making `alist` more or less pointless.

Would such a combinator help in your scenario?

**Reply by ArtemyB** [reply]
Wow, thanks for such a quick reaction 👍 

> the main problem here is, that the int may actually change when something is inserted into the list.

Got it, that's what I suspected.

> It would however be possible to implement something like AList.mapIntIndex : (aval<int> -> 'a -> 'b) -> alist<'a> -> alist<'b> where the aval<int> will change appropriately.

Would it be possible via that function to construct `AMap<Index, int>`, which can be further used to retrieve corresponding `int`-index?

**Reply by ArtemyB** [reply]
>  Of course for small lists you can always do something like list |> AList.toAVal |> AVal.map (Seq.mapi mapping) |> AList.ofAVal but that looses adaptivity on its elements making alist more or less pointless.

Hmm, this is a good option just to make the thing done -- luckily I don't have big lists currently, so this should work fine.

I think I don't require adaptivity in this case (to retrieve an `int`-index). For clarification, here is what I'm solving:

- There is a spreadsheet-like UI component that is used to output a set of data. The data can be changed (by the component itself and from the outside). Before output to the component the data could also be preprocessed (e.g. filtered).
- For rendering the component uses a callback function of kind `columnIndex: int -> rowIndex: int -> CellView`. The callback is passed to the component on its initialization as a parameter.
- When some cell should be re-rendered, I have to notify the component to do so by specifying the `int`-coordinates of the updated cell -- that's when I have to know the `int` (so seems like it doesn't need to be adaptive). And the notification call itself is handled via `AddCallback` on the output adaptive list.

**Comment by krauthaufen** [comment, marked as answer]
So in essence you just need the int in a callback to communicate with the outside world? This should be fairly simple to achieve with reasonable runtime (e.g. less than linear in the number of elements). I can put something together next week if i don't forget it....

**Reply by ArtemyB** [reply]
> So in essence you just need the int in a callback to communicate with the outside world

Actually right, my current solution could be reduced to that 🤔 (I'm just using intermediate events for the sake of API intuitiveness, but it doesn't change the essence)

**Reply by ArtemyB** [reply]
And big thanks for the help and for the great library itself -- it seems perfect for the task 👍

**Reply by ArtemyB** [reply]
Btw I can create an issue based on this discussion if that could be useful somehow 🤷‍♂️
